### PR TITLE
Add SEP-991 (CIMD) support for URL-based client IDs

### PR DESCRIPTION
**Pull request by tanish111**
- Added SEP-991 ([CIMD](https://oauth.net/2/client-id-metadata-document/)) support in crates/rmcp/src/transport/auth.rs. 
- Updated AuthorizationSession::new() to accept an optional client_metadata_url and check client_id_metadata_document_supported in the authorization server metadata. When a valid metadata URL is provided and supported, the client uses the URL as the client_id (CIMD flow); otherwise, it falls back to dynamic client registration.
- Added OAuthState::start_authorization_with_metadata_url(), and updated start_authorization() to call it with None for backward compatibility. This maintains existing behavior while enabling URL-based client IDs when servers support CIMD.

## Motivation and Context
Implement SEP-991: URL-based Client Registration (OAuth Client ID Metadata)

## How Has This Been Tested?

1. Created a CIMD-compliant OAuth server example (`cimd_auth_streamhttp.rs`) implementing SEP-991: validates client_id URLs, fetches and validates client metadata documents, and enforces CIMD restrictions according to [specs](https://oauth.net/2/client-id-metadata-document/). This is tested against external CIMD-compliant testing tools ([client.dev](https://client.dev/) and [example-app.com/client](https://oauth.net/2/client-id-metadata-document/)). 
Testing Results form client.dev for example server:- 

Screenshot 2025-12-05 at 12 44 57 AM

3. Updated oauth_client.rs to support CIMD with 2 CLI arguments (server URL and client metadata URL). End-to-end flow verified: 
Start the server using `cargo run --example servers_cimd_auth_streamhttp`
In a seperate terminal run the client with `cargo run --example clients_oauth_client --package mcp-client-examples -- http://127.0.0.1:3000 https://raw.githubusercontent.com/tanish111/cimd-local-oauth-server/refs/heads/main/client-metadata.json`. And complete the OAuth Flow.

4. Added unit tests for the is_https_url helper covering valid HTTPS URLs, invalid schemes, root paths, and edge cases. 

## Breaking Changes
No

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist

- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [ ] I have added or updated documentation as needed

## Additional context
Fix issue #519

### CODE REVIEW

File: crates/rmcp/src/transport/auth.rs [hunk 1]
```diff
@@ -1256,7 +1316,50 @@ impl OAuthState {
 mod tests {
     use url::Url;
 
-    use super::AuthorizationManager;
+    use super::{AuthorizationManager, is_https_url};
+
+    // SEP-991: URL-based Client IDs
```
Comment (jokemanfire) [inline crates/rmcp/src/transport/auth.rs:1315]: merge to one test pls.

**Comment by jokemanfire** [comment]
There's a interface set_credentials , Can it already cover CIMD？

File: crates/rmcp/src/transport/auth.rs [hunk 2]
```diff
@@ -239,6 +239,15 @@ struct AuthorizationState {
...
 }
 
+/// SEP-991: URL-based Client IDs
+/// Validate that the client_id is a valid URL with https scheme and non-root pathname
+fn is_https_url(value: &str) -> bool {
+    match Url::parse(value) {
+        Ok(url) => url.scheme() == "https" && url.path() != "/",
+        Err(_) => false,
+    }
+}
```
Comment (jokemanfire) [inline crates/rmcp/src/transport/auth.rs:249]: fn is_https_url(value: &str) -> bool {
    Url::parse(value)
        .ok()
        .map(|url| url.scheme() == "https" && url.path() != "/" && url.host_str().is_some())
        .unwrap_or(false)
}

File: crates/rmcp/src/transport/auth.rs [hunk 3]
```diff
@@ -1125,6 +1167,18 @@ impl OAuthState {
         scopes: &[&str],
         redirect_uri: &str,
         client_name: Option<&str>,
+    ) -> Result<(), AuthError> {
```
Comment (jokemanfire) [inline crates/rmcp/src/transport/auth.rs:1164]: I think use set_credentials interface is enough ?

Comment (tanish111) [reply crates/rmcp/src/transport/auth.rs:1164]: In that case, I’ll need to update the `start_authorization` function signature to accept `client_metadata_url `as an optional parameter. This change will break existing MCP servers build using `start_authorization`, since they will now need to pass None as the final argument.

File: crates/rmcp/src/transport/auth.rs (hunk 1 above)
Comment (tanish111) [reply crates/rmcp/src/transport/auth.rs:1315]: Yes I would do this.

File: crates/rmcp/src/transport/auth.rs (hunk 2 above)
Comment (tanish111) [reply crates/rmcp/src/transport/auth.rs:249]: I will update this.
Old algorithm was missing `https://` .

File: crates/rmcp/src/transport/auth.rs (hunk 1 above)
Comment (tanish111) [reply crates/rmcp/src/transport/auth.rs:1315]: Fixed.

File: crates/rmcp/src/transport/auth.rs (hunk 2 above)
Comment (tanish111) [reply crates/rmcp/src/transport/auth.rs:249]: Fixed.

File: crates/rmcp/src/transport/auth.rs (hunk 3 above)
Comment (tanish111) [reply crates/rmcp/src/transport/auth.rs:1164]: @jokemanfire another drawback is `set_credentials` only injects existing tokens; it cannot run the auth flow to obtain them, so we still need `start_authorization*`

File: crates/rmcp/src/transport/auth.rs [hunk 4]
```diff
@@ -950,30 +959,63 @@ impl AuthorizationSession {
...
+                    client_secret: None,
+                    scopes: scopes.iter().map(|s| s.to_string()).collect(),
+                    redirect_uri: redirect_uri.to_string(),
+                }
+            } else {
+                // Fallback to dynamic registration
+                match auth_manager
+                    .register_client(client_name.unwrap_or("MCP Client"), redirect_uri)
+                    .await
+                {
```
Comment (jokemanfire) [inline crates/rmcp/src/transport/auth.rs]: map_error?

**Comment by jokemanfire** [comment]
LGTM ,but a little grammar simplification.

File: crates/rmcp/src/transport/auth.rs (hunk 4 above)
Comment (tanish111) [reply crates/rmcp/src/transport/auth.rs]: @jokemanfire fixed.

**Comment by tanish111** [comment]
@alexhancock can you also review it?

**Review by jokemanfire — DISMISSED** [review]

File: examples/clients/src/auth/oauth_client.rs [hunk 5]
```diff
@@ -23,10 +23,11 @@ use tokio::{
 };
 use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt};
 
-const MCP_SERVER_URL: &str = "http://localhost:3000/mcp";
-const MCP_REDIRECT_URI: &str = "http://localhost:8080/callback";
+const MCP_SERVER_URL: &str = "http://127.0.0.1:3000/mcp";
+const MCP_REDIRECT_URI: &str = "http://127.0.0.1:8080/callback";
 const CALLBACK_PORT: u16 = 8080;
 const CALLBACK_HTML: &str = include_str!("callback.html");
+const CLIENT_METADATA_URL: &str = "https://raw.githubusercontent.com/tanish111/cimd-local-oauth-server/refs/heads/main/client-metadata.json";
```
Comment (alexhancock) [inline examples/clients/src/auth/oauth_client.rs]: can you put one in the rust-sdk repo itself?

**Review by alexhancock — DISMISSED** [review]

File: examples/clients/src/auth/oauth_client.rs (hunk 5 above)
Comment (tanish111) [reply examples/clients/src/auth/oauth_client.rs]: @alexhancock yes I can

Comment (tanish111) [reply examples/clients/src/auth/oauth_client.rs]: @alexhancock I have updated it with new metadata json hosted on rust-sdk repo.
